### ISSUE TITLE
build failure fetching from internal mirror

### ISSUE THREAD

**Issue by grwilson** [delphix/appliance-build#380]
We have seen several builds fail with the following error:

```
00:34:40  Get:641 http://linux-package-mirror.****.com/master/415f972b/ppas bionic-updates/main amd64 openssh-server-dbgsym amd64 1:7.6p1-4ubuntu0.3 [884 kB]
00:34:40  Get:642 http://linux-package-mirror.****.com/master/415f972b/ppas bionic-updates/main amd64 systemd-dbgsym amd64 237-3ubuntu10.29 [7677 kB]
00:34:40  Fetched 1379 MB in 18s (77.6 MB/s)
00:34:40  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
00:34:40  E: Failed to fetch http://linux-package-mirror.****.com/master/415f972b/ubuntu/pool/universe/t/trace-cmd/trace-cmd_2.6.1-0.1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
00:34:40  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```

Looking earlier in the build we see the `acl` package was successfully accessed:
```00:27:21  Get:221 http://linux-package-mirror.****.com/master/415f972b/ubuntu bionic/main amd64 acl amd64 2.2.52-3build1 [38.5 kB]
00:27:40  Selecting previously unselected package acl.
00:27:40  Preparing to unpack .../126-acl_2.2.52-3build1_amd64.deb ...
00:27:40  Unpacking acl (2.2.52-3build1) ..

00:28:32  Setting up acl (2.2.52-3build1) ...
```

And as we start to run the upgrade build image we see the following:
```
00:34:22  Get:1 http://linux-package-mirror.****.com/master/415f972b/ubuntu bionic/main amd64 acl amd64 2.2.52-3build1 [38.5 kB]
```

But it fails later:
```
00:34:40  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
00:34:40  E: Failed to fetch http://linux-package-mirror.****.com/master/415f972b/ubuntu/pool/universe/t/trace-cmd/trace-cmd_2.6.1-0.1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
00:34:40  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```

**Comment by grwilson** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/master/job/nightly/268/

**Comment by grwilson** [comment]
It's interesting that the failure always seems to happen after we get this warning:

**Comment by grwilson** [comment]
Back to the original example, looking closer at the failure of `trace-cmd`:

1. We first download it successfully here:
```00:27:24  Get:410 http://linux-package-mirror.****.com/master/415f972b/ubuntu bionic/universe amd64 trace-cmd amd64 2.6.1-0.1 [298 kB]
```
2. Unpack it:
```00:28:18  Selecting previously unselected package trace-cmd.
00:28:18  Preparing to unpack .../349-trace-cmd_2.6.1-0.1_amd64.deb ...
00:28:18  Unpacking trace-cmd (2.6.1-0.1) ...```
3. Later we start to create the upgrade image:
```00:34:21  + xargs apt-get download
00:34:21  + dpkg-query -Wf '${Package}=${Version}\n'
00:34:24  Get:297 http://linux-package-mirror.****.com/master/415f972b/ubuntu bionic/universe amd64 trace-cmd amd64 2.6.1-0.1 [298 kB]
00:34:24  Err:297 http://linux-package-mirror.****.com/master/415f972b/ubuntu bionic/universe amd64 trace-cmd amd64 2.6.1-0.1
00:34:24    Undetermined Error [IP: 10.110.20.241 80]
```
4. Then later we fail with the error:
```00:34:40  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
00:34:40  E: Failed to fetch http://linux-package-mirror.****.com/master/415f972b/ubuntu/pool/universe/t/trace-cmd/trace-cmd_2.6.1-0.1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]```

**Comment by grwilson** [comment]
Is it possible that the mirror is getting overloaded with all the builds trying to fetch packages at the same time? Should we retry the ` dpkg-query -Wf '${Package}=${Version}\n' | xargs apt-get download` command on failure?

**Comment by prashks** [comment]
Another instance: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/1057/console

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/21974/console

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/patch/job/post-push/436/console

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/1131/console

**Comment by prashks** [comment]
One more: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/1171/console

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/674/console

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/732/console

**Comment by prashks** [comment]
Another: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/22335/console
```05:52:38  Fetched 304 MB in 4s (79.9 MB/s)
05:52:38  E: Failed to fetch http://linux-package-mirror.****.com/master/415f972b/ubuntu/pool/main/r/rtmpdump/librtmp1_2.4+20151223.gitfa8646d.1-1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
05:52:38  E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

**Comment by prashks** [comment]
Again: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/22677/console

**Comment by sdimitro** [comment]
Another instance: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/1014/execution/node/81/log/

**Comment by sdimitro** [comment]
Hit again:
```
06:46:16  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
06:46:16  E: Failed to fetch http://linux-package-mirror.****.com/6.0/stage/66c6fa02/ubuntu/pool/main/libt/libthai/libthai-data_0.1.27-2_all.deb  Undetermined Error [IP: 10.110.20.241 80]
06:46:16  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/1099/execution/node/81/log/

**Comment by sdimitro** [comment]
Hit again:
```
01:47:40  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
01:47:40  E: Failed to fetch http://linux-package-mirror.****.com/6.0/release/f3caf3e6/ubuntu/pool/main/libt/libthai/libthai-data_0.1.27-2_all.deb  Undetermined Error [IP: 10.110.20.241 80]
01:47:40  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2028/execution/node/81/log/

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2055/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/release/job/post-push/82/

**Comment by prakashsurya** [comment]
```
E: Failed to fetch http://linux-package-mirror.****.com/master/6e095cbd/ubuntu/pool/main/x/x11-utils/x11-utils_7.7+3build1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n", "stderr_lines": ["E: Failed to fetch http://linux-package-mirror.****.com/master/6e095cbd/ubuntu/pool/main/x/x11-utils/x11-utils_7.7+3build1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]", "E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?"
```

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/23564/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2402/console

**Comment by prakashsurya** [comment]
```
00:39:04.332  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
00:39:04.332  E: Failed to fetch http://linux-package-mirror.****.com/6.0/stage/66c6fa02/ubuntu/pool/main/libt/libthai/libthai-data_0.1.27-2_all.deb  Undetermined Error [IP: 10.110.20.241 80]
00:39:04.332  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/1336/console

**Comment by prakashsurya** [comment]
```
00:45:40.613  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
00:45:40.613  E: Failed to fetch http://linux-package-mirror.****.com/6.0/patch/13f0472d/ubuntu/pool/main/p/python-crypto/python3-crypto_2.6.1-8ubuntu2_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
00:45:40.613  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/patch/job/post-push/785/console

**Comment by pzakha** [comment]
```
01:50:20  E: Failed to fetch http://linux-package-mirror.****.com/6.0/release/f3caf3e6/ubuntu/pool/main/libt/libthai/libthai-data_0.1.27-2_all.deb  Undetermined Error [IP: 10.110.20.241 80]
01:50:20  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
```
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2705/console

**Comment by pzakha** [comment]
More failures.

Failed downloading debs in `81-upgrade-repository.binary`:
- http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2801/console
- http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/2833/console
- http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/1601/console

Failed building image in `80-build-configuration.binary`:
- http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/24025/console

**Comment by sebroy** [comment]
The internal mirror has datadog monitoring here: https://app.datadoghq.com/dash/host/910199161?live=4h&page=0&tile_size=m

We should take a look at this to understand if any resource utilization issues exist on the mirror when these problems occur.

**Comment by pzakha** [comment]
More failures, most of them with `libthai-data`, but I've also seen `acl` fail from time to time:
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/patch/job/post-push/1058/console

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/3830/consoleFull

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/2048/console

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/3797/consoleFull

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/2023/console

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/3645/console

**Comment by shartse** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/3613/console

These all have been a mix of `libthai` and `acl`

**Comment by don-brady** [comment]
more with `libthai`...

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/nightly/59/

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/nightly/60/

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/release/job/nightly/62/

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/post-push/87/

**Comment by jwk404** [comment]
acl_2.2.52-3build1_amd64.deb:

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/4252/console

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/stage/job/post-push/2305/console

**Comment by jwk404** [comment]
acl_2.2.52-3build1_amd64.deb:

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/patch/job/nightly/69/
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/nightly/67/
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/release/job/nightly/68/

**Comment by jwk404** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/25670/console

```
23:21:12  E: config/hooks/80-build-configuration.binary failed (exit non-zero). You should check for errors.
23:21:12  P: Begin unmounting filesystems...
23:21:12  P: Saving caches...
23:21:12  Reading package lists...
23:21:12  Building dependency tree...+ kill -9 312
23:21:12  + [[ ! -f binary/SHA256SUMS ]]
23:21:12  + exit 1
```

**Comment by jwk404** [comment]
acl_2.2.52-3build1_amd64.deb:

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/patch/job/nightly/72/console
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/post-push/98/console

**Comment by pzakha** [comment]
Another failure, this time in `80-build-configuration.binary` when installing packages required by the `appliance-build.unittest-internal` ansible role.
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/27042/consoleFull
```
fatal: [binary]: FAILED! => {"cache_update_time": 1576823027, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"      install 'adoptopenjdk-java8-jdk' 'ant' 'build-essential' 'docker.io' 'git' 'chromium-browser'' failed: E: Failed to fetch http://linux-package-mirror.****.com/master/793d3297/ubuntu/pool/main/x/x11-utils/x11-utils_7.7+3build1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n", "rc": 100, "stderr": "E: Failed to fetch http://linux-package-mirror.****.com/master/793d3297/ubuntu/pool/main/x/x11-utils/x11-utils_7.7+3build1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n"
```

**Comment by pzakha** [comment]
There was more than 20 failures spread out over 5 builds during Christmas, most of them while updating packages in the docker container. 
e.g.: http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/patch/job/nightly/92/

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/29099/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/29000/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/6648/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/6799/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/6916/console

**Comment by prakashsurya** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/release/job/post-push/7117/console

**Comment by pcd1193182** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/patch/job/post-push/3322/console

**Comment by don-brady** [comment]
http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/29843/consoleFull

**Comment by jwk404** [comment]
Two failures from http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build/job/6.0/job/stage/job/nightly/126/console

```
23:24:40  E: Failed to fetch http://linux-package-mirror.****.com/6.0/stage/c5e9be29/ubuntu/pool/main/g/gtk+2.0/libgtk2.0-0_2.24.32-1ubuntu1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
23:24:40  E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
23:24:40  P: Begin unmounting filesystems...
```

and

```
23:24:32  E: Failed to fetch http://linux-package-mirror.****.com/6.0/stage/c5e9be29/ubuntu/pool/main/o/os-prober/os-prober_1.74ubuntu1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
23:24:32  E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
23:24:32  P: Begin unmounting filesystems...
```

**Comment by jwk404** [comment]
Two more:

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/6.0/job/patch/job/post-push/3620/console

```
23:24:47  W: Download is performed unsandboxed as root as file '/packages/acl_2.2.52-3build1_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
23:24:47  E: Failed to fetch http://linux-package-mirror.****.com/6.0/patch/13f0472d/ubuntu/pool/universe/f/fio/fio_3.1-1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
23:24:47  E: config/hooks/81-upgrade-repository.binary failed (exit non-zero). You should check for errors.
23:24:47  P: Begin unmounting filesystems...
```

http://ops.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-stage1/job/master/job/post-push/30447/console

```
01:13:42  E: Failed to fetch http://linux-package-mirror.****.com/master/38b5a659/ubuntu/pool/main/i/iproute2/iproute2_4.15.0-2ubuntu1_amd64.deb  Undetermined Error [IP: 10.110.20.241 80]
01:13:42  E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
01:13:42  P: Begin unmounting filesystems...
```

**Comment by pzakha** [comment]
This got close because I've added a "fixes" field to #415. With that PR pushed we should hopefully not see this issue anymore on `master`. Once we have enough confidence that this has indeed been fixed we can backport this PR to other branches.
